api/client/request.go

```diff
@@ -0,0 +1,84 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptrace"
+	"time"
+
+	"github.com/brimdata/zed/api"
+)
+
+type Request struct {
+	Header  http.Header
+	Method  string
+	Path    string
+	Body    interface{}
+	RawBody io.Reader
+
+	host string
+	ctx  context.Context
+
+	// trace fields
+	dnsStartTime  time.Time
+	firstByteTime time.Time
+	getConnTime   time.Time
+	gotConnInfo   httptrace.GotConnInfo
+}
+
+func newRequest(ctx context.Context, host string, h http.Header) *Request {
+	if requestID := api.RequestIDFromContext(ctx); requestID != "" {
+		h.Set(api.RequestIDHeader, requestID)
+	}
+	req := &Request{
+		Header: h,
+		host:   host,
+	}
+	// use trace to track timing
+	req.ctx = httptrace.WithClientTrace(ctx, &httptrace.ClientTrace{
+		DNSStart:             func(httptrace.DNSStartInfo) { req.dnsStartTime = time.Now() },
+		GotConn:              func(g httptrace.GotConnInfo) { req.gotConnInfo = g },
+		GetConn:              func(string) { req.getConnTime = time.Now() },
+		GotFirstResponseByte: func() { req.firstByteTime = time.Now() },
+	})
+	return req
+}
+
+func (r *Request) reader() (io.Reader, error) {
```

Comment:
NIt: This is less interesting than HTTPRequest, so I'd put it after that method.